### PR TITLE
Sync off-chain to blockchain

### PR DESCRIPTION
**Pull request by LiorRabin**
## Proposed changes

This PR handles the synchronisation of off-chain transactions to the blockchain.
A kinda big refactor was made in order for this to work and address all issues related to the process (and still more need to be handled in separate PRs).

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/ColuLocalNetwork/inventory-manager/blob/master/.github/CONTRIBUTING.md) doc
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

Need to review this PR carefully :)

### CODE REVIEW

File: modules/lib/models/transaction.js [hunk 1]
```diff
@@ -16,24 +19,326 @@ module.exports = (osseus) => {
     })
   }
 
-  transaction.create = (from, to, amount) => {
+  const validateAmount = (amount) => {
+    return new Promise(async (resolve, reject) => {
+      amount = new BigNumber(amount)
+      if (amount.eq(0) || amount.lt(0)) {
```
Comment (bejavu) [inline modules/lib/models/transaction.js]: you can use lte, or not gt
but why do we use this validate amount at all?

File: modules/lib/models/transaction.js [hunk 2]
```diff
@@ -16,24 +19,326 @@ module.exports = (osseus) => {
...
+        const tasks = []
+        const wallets = await osseus.db_models.wallet.getAll()
+        osseus.logger.debug(`validateBlockchainBalances - found ${wallets.length} wallets`)
+        wallets.forEach(wallet => {
+          let address = wallet.address
+          let currencies = wallet.balances.map(balance => balance.currency.ccAddress)
+          currencies.forEach(currency => {
+            tasks.push(new Promise(async (resolve, reject) => {
+              try {
+                let valid = await osseus.utils.validateBlockchainBalance(address, currency)
```
Comment (bejavu) [inline modules/lib/models/transaction.js]: i don't think that a call to a logic from the utils should be done from the model class
controllers (or util logic) should call the model functions, not the other way around

File: modules/lib/models/transaction.js [hunk 3]
```diff
@@ -16,24 +19,326 @@ module.exports = (osseus) => {
...
+      } catch (err) {
+        reject(err)
+      }
+    })
+  }
+
+  const validateAggregatedBalances = () => {
+    return new Promise(async (resolve, reject) => {
+      try {
+        const results = await osseus.utils.validateAggregatedBalances()
```
Comment (bejavu) [inline modules/lib/models/transaction.js]: same here

File: modules/lib/models/transaction.js [hunk 4]
```diff
@@ -16,24 +19,326 @@ module.exports = (osseus) => {
...
+        const tasks = []
+        const wallets = await osseus.db_models.wallet.getAll()
+        osseus.logger.debug(`updateBlockchainBalances - found ${wallets.length} wallets`)
+        wallets.forEach(wallet => {
+          let address = wallet.address
+          let currencies = wallet.balances.map(balance => balance.currency.ccAddress)
+          currencies.forEach(currency => {
+            tasks.push(new Promise(async (resolve, reject) => {
+              try {
+                let updated = await osseus.utils.updateBlockchainBalance(address, currency)
```
Comment (bejavu) [inline modules/lib/models/transaction.js]: and here
you should move this logic to the controller \ utils

File: modules/lib/models/transaction.js [hunk 5]
```diff
@@ -16,24 +19,326 @@ module.exports = (osseus) => {
...
+        reject(err)
+      }
+    })
+  }
+
+  const transmitToBlockchain = (transmit, bctxs, txids) => {
+    return new Promise(async (resolve, reject) => {
+      try {
+        const transmittedBctxs = await Promise.mapSeries(bctxs, async tx => {
+          let bctx = await osseus.lib.BlockchainTransaction.transfer(tx.from, tx.to, tx.token, tx.amount, tx.opts)
```
Comment (bejavu) [inline modules/lib/models/transaction.js]: also here
the model should not "know" the blockchain.

File: modules/lib/models/transaction.js (hunk 2 above)
Comment (LiorRabin) [reply modules/lib/models/transaction.js]: it's in the `lib` so all is ok :)

**Pull request merged**